AnimationGroup.lua

```diff
@@ -60,28 +60,50 @@ function CreateFrame(...)
     return frame
 end
 
-
 local function OnUpdate(self, elapsed)
     if self.paused then
         return
     end
+    local in_reverse = self.group.reverse
+    local is_start_delayed = self.startDelay and self.startDelayTime < self.startDelay
+    local is_end_delayed = self.endDelay and self.endDelayTime < self.endDelay
```

Comment:
According to this:

> * Animation:SetStartDelay(delaySec)
>   Set the number of seconds that the animation delays before it starts to progress.
> * Animation:SetEndDelay(delaySec)
>   Set the number of seconds the animation delays after finishing.

Meaning,

```
 <-----------------------------------> total_time
              <-----------> self.duration
 | startDelay | animation | endDelay |
 min                               max
          ^
 <--------|-------------------------->
      self.time
        
```

which simplifies this a great deal:
```lua
local total_time = (self.startDelay or 0) + self.duration + (self.endDelay or 0)
self.time = self.time + (self.group.reverse and -elapsed or elapsed)
local animation_time = self.time - (self.startDelay or 0) - (self.endDelay or 0)

self.progress = animation_time / self.duration
-- Fast min/max to ensure progress is bound within `0 < self.progress < 1'
self.progress = self.progress < 0 and 0 or self.progress
self.progress = self.progress > 1 and 1 or self.progress

-- These represent whether the time "pointer" is in the left- or the right delay part, at which it is not animating, but should remain visible.
local is_delayed_start = self.time < self.startDelay or 0 and 0 < self.time
local is_delayed_end = (total_time - self.endDelay or 0) < self.time and self.time < total_time

local is_animation_complete = (in_reverse and self.time < 0) or total_time < self.time
local is_animating = (self.startDelay or 0) < self.time and self.time < total_time - (self.endDelay or 0)
self.delayed = is_delayed_start or is_delayed_end

if is_animation_complete and not self.delayed then
   AG:Stop(self)
   AG:Fire(self.group, self, 'Finished')
   return
end

-- ...
```

`animation_time` may be negative if delays are configured, but that is okay, since it just means the animation hasn't begun.

Reply:
yea may be its good solution. Ill test it and look for other parts....
there are GetElapsed  - return elapsed time include delays... so mb we need name self.time = self.elapsed?

Reply:
I agree `self.time` is not a good name for what it's all about, never was. `self.timeProgress` would make more sense - it's almost the actual progress, but it spans the full time duration incl. delays.

Reply:
Also, GetElapsed,

> * Animation:GetElapsed()
>  Gets the amount of time in seconds that the animation has been playing for.

So that would almost be `animation_time` accounting for `in_reverse`, ensuring it's considered from which end the animation starts.